File: webdisplay/backend/data_processor.py
"""
Data processing module for flight test data.
Handles CSV loading, quaternion calculations, and frame matching.
"""
import os
import pandas as pd
import math
from pathlib import Path
from bisect import bisect_left
from scipy.spatial.transform import Rotation

class FlightDataProcessor:
    """Processes flight test data and provides access methods"""
    
    def __init__(self, data_file, frames_directory=None):
        """
        Initialize processor with data file.
        
        Args:
            data_file: Path to merged_data.csv
            frames_directory: Optional - not used anymore (kept for backward compatibility)
        """
        self.data_file = Path(data_file)
        
        # Load and process data
        self._load_data()
        
    def _load_data(self):
        """Load and process the merged_data.csv file"""
        print(f"Loading data from {self.data_file}...")
        self.merged_data = pd.read_csv(self.data_file)
        
        # Robust timestamp parsing
        ts_as_td = pd.to_timedelta(self.merged_data['timestamp'], errors='coerce')
        if ts_as_td.isna().all():
            # Fallback: treat as numeric seconds
            sec = pd.to_numeric(self.merged_data['timestamp'], errors='coerce')
            ts_as_td = pd.to_timedelta(sec, unit='s', errors='coerce')
        
        self.merged_data['timestamp'] = ts_as_td
        self.merged_data = self.merged_data.dropna(subset=['timestamp'])
        self.merged_data.set_index('timestamp', inplace=True)
        
        # Convert to list for easier access
        self.data_list = list(self.merged_data.iterrows())
        
        # CRITICAL: Pre-compute numeric timestamps for fast binary search
        # This avoids converting timedelta to seconds on every lookup
        self.timestamp_ns_list = [idx.value for idx, _ in self.data_list]
        
        print(f"Loaded {len(self.data_list)} data rows")
        
    def get_summary(self):
        """Get summary of loaded data"""
        return {
            'data_count': len(self.data_list),
            'data_columns': list(self.merged_data.columns)
        }
    
    def get_data_count(self):
        """Get total number of data points"""
        return len(self.data_list)
    
    def get_data_at_index(self, index):
        """
        Get flight data at a specific index.
        Returns dict with all relevant data including quaternions.
        """
        if index < 0 or index >= len(self.data_list):
            return None
        
        idx, row = self.data_list[index]
        timestamp = idx  # pandas Timedelta
        
        # Calculate quaternions (same logic as playback.py)
        helmet_corrected = Rotation.from_quat([
            row['x_helmet'], row['y_helmet'], row['z_helmet'], row['w_helmet']
        ])
        vehicle_world = Rotation.from_quat([
            row['x_vehicle'], row['y_vehicle'], row['z_vehicle'], row['w_vehicle']
        ])
        helmet_world = vehicle_world * helmet_corrected
        
        vqx, vqy, vqz, vqw = vehicle_world.as_quat()
        hqx, hqy, hqz, hqw = helmet_world.as_quat()
        
        # Calculate ground speed
        vvn = row['north']
        vve = row['east']
        vvd = row['down']
        gspeed = math.sqrt(vvn**2 + vve**2 + vvd**2)
        
        return {
            'index': index,
            'timestamp_seconds': timestamp.total_seconds(),
            'timestamp_ns': timestamp.value,
            
            # Vehicle quaternion (world-relative)
            'VQX': float(vqx),
            'VQY': float(vqy),
            'VQZ': float(vqz),
            'VQW': float(vqw),
            
            # Helmet quaternion (world-relative)
            'HQX': float(hqx),
            'HQY': float(hqy),
            'HQZ': float(hqz),
            'HQW': float(hqw),
            
            # Vehicle position and velocity
            'VLAT': float(row['lat']),
            'VLON': float(row['lon']),
            'VALT': float(row['alt']),
            'VVN': float(vvn),
            'VVE': float(vve),
            'VVD': float(vvd),
            'VINS': int(row['mode']) if 'mode' in row else 0,
            'GSPEED': float(gspeed),
        }
    
    def find_index_for_timestamp(self, timestamp_seconds):
        """
        Find the data index closest to a given timestamp (in seconds).
        Uses binary search for O(log n) performance instead of O(n) linear search.
        """
        if not self.data_list:
            return None
        
        # Ensure timestamp_ns_list exists (for backward compatibility with old instances)
        if not hasattr(self, 'timestamp_ns_list') or not self.timestamp_ns_list:
            # Fallback: create it now if it doesn't exist
            self.timestamp_ns_list = [idx.value for idx, _ in self.data_list]
        
        # Convert target timestamp to nanoseconds for comparison
        target_td = pd.to_timedelta(timestamp_seconds, unit='s')
        target_ns = target_td.value
        
        # Binary search for closest timestamp
        # bisect_left finds the insertion point, which gives us the closest index
        idx = bisect_left(self.timestamp_ns_list, target_ns)
        
        # Handle edge cases
        if idx == 0:
            return 0
        elif idx >= len(self.timestamp_ns_list):
            return len(self.timestamp_ns_list) - 1
        
        # Check which of the two adjacent points is closer
        prev_diff = abs(self.timestamp_ns_list[idx - 1] - target_ns)
        next_diff = abs(self.timestamp_ns_list[idx] - target_ns)
        
        if prev_diff < next_diff:
            return idx - 1
        else:
            return idx
    
    def get_all_path_data(self):
        """
        Efficiently extract all lat/lon/alt data (like playback.py lines 202-204).
        Returns dict with 'lats', 'lons', 'alts' lists.
        """
        # Fast list comprehension over already-loaded data
        lats = [row['lat'] for _, row in self.data_list]
        lons = [row['lon'] for _, row in self.data_list]
        alts = [row['alt'] for _, row in self.data_list]
        
        return {
            'lats': lats,
            'lons': lons,
            'alts': alts
        }
    
    def get_all_attitude_data(self):
        """
        Efficiently extract all attitude data (yaw, pitch, roll) from vehicle quaternions.
        Returns dict with 'yaws', 'pitches', 'rolls' lists (in degrees).
        """
        yaws = []
        pitches = []
        rolls = []
        
        for _, row in self.data_list:
            # Calculate vehicle quaternion (same as get_data_at_index)
            vehicle_world = Rotation.from_quat([
                row['x_vehicle'], row['y_vehicle'], row['z_vehicle'], row['w_vehicle']
            ])
            
            # Convert quaternion to Euler angles (in radians)
            # scipy uses 'xyz' convention: roll (x), pitch (y), yaw (z)
            euler = vehicle_world.as_euler('xyz', degrees=False)
            
            # Extract and convert to degrees
            roll = math.degrees(euler[0])   # Roll (x-axis rotation)
            pitch = math.degrees(euler[1])   # Pitch (y-axis rotation)
            yaw = math.degrees(euler[2])     # Yaw (z-axis rotation)
            
            pitches.append(pitch)
            rolls.append(roll)
            yaws.append(yaw)
        
        return {
            'yaws': yaws,
            'pitches': pitches,
            'rolls': rolls
        }

File: webdisplay/backend/dataProcessor.js
/**
 * Data processing module for flight test data.
 * Handles CSV loading, quaternion calculations, and frame matching.
 */

const fs = require('fs');
const path = require('path');
const { parse } = require('csv-parse/sync');
const { quat, mat4 } = require('gl-matrix');

class FlightDataProcessor {
    /**
     * Initialize processor with data file.
     * @param {string} dataFile - Path to merged_data.csv
     */
    constructor(dataFile) {
        this.dataFile = dataFile;
        this.dataList = [];
        this.timestampNsList = [];
        this._loadData();
    }

    /**
     * Load and process the merged_data.csv file
     */
    _loadData() {
        console.log(`[DataProcessor] Loading data from ${this.dataFile}...`);
        
        // Check if file exists
        if (!fs.existsSync(this.dataFile)) {
            console.error(`[DataProcessor] ERROR: Data file not found: ${this.dataFile}`);
            this.dataList = [];
            return;
        }
        
        const fileContent = fs.readFileSync(this.dataFile, 'utf-8');
        console.log(`[DataProcessor] File size: ${fileContent.length} bytes`);
        
        const records = parse(fileContent, {
            columns: true,
            skip_empty_lines: true,
            cast: (value, context) => {
                // Don't cast timestamp column - it needs special parsing
                if (context.column === 'timestamp') {
                    return value; // Keep as string for timestamp parsing
                }
                // Try to parse as number for other columns
                const num = parseFloat(value);
                if (!isNaN(num) && isFinite(num)) {
                    return num;
                }
                return value;
            }
        });

        // Only log errors
        if (records.length === 0) {
            console.error(`[DataProcessor] ERROR: No records parsed from CSV file!`);
        } else {
            const firstRecord = records[0];
            const columns = Object.keys(firstRecord);
            const requiredColumns = ['lat', 'lon', 'alt', 'x_vehicle', 'y_vehicle', 'z_vehicle', 'w_vehicle'];
            const missingColumns = requiredColumns.filter(col => !columns.includes(col));
            if (missingColumns.length > 0) {
                console.error(`[DataProcessor] ERROR: Missing required columns: ${missingColumns.join(', ')}`);
                console.error(`[DataProcessor] Available columns: ${columns.join(', ')}`);
            }
        }

        // Process timestamps and convert to seconds
        this.dataList = records.map((row, index) => {
            // Parse timestamp - handle both timedelta strings and numeric seconds
            let timestampSeconds = 0;
            if (typeof row.timestamp === 'string') {
                // Try parsing as timedelta (e.g., "0 days 00:00:02.643000")
                const tdMatch = row.timestamp.match(/(\d+) days?\s+(\d+):(\d+):([\d.]+)/);
                if (tdMatch) {
                    const days = parseInt(tdMatch[1]) || 0;
                    const hours = parseInt(tdMatch[2]) || 0;
                    const minutes = parseInt(tdMatch[3]) || 0;
                    const seconds = parseFloat(tdMatch[4]) || 0;
                    timestampSeconds = days * 86400 + hours * 3600 + minutes * 60 + seconds;
                } else {
                    // Try as numeric seconds
                    timestampSeconds = parseFloat(row.timestamp) || 0;
                }
            } else {
                timestampSeconds = parseFloat(row.timestamp) || 0;
            }

            return {
                index,
                timestampSeconds,
                timestampNs: timestampSeconds * 1e9, // Convert to nanoseconds for binary search
                row
            };
        });

        // Sort by timestamp
        this.dataList.sort((a, b) => a.timestampSeconds - b.timestampSeconds);

        // Pre-compute numeric timestamps for fast binary search
        this.timestampNsList = this.dataList.map(d => d.timestampNs);

        if (this.dataList.length === 0) {
            console.error(`[DataProcessor] ERROR: No data rows loaded! Check CSV file format.`);
        }
    }

    /**
     * Get summary of loaded data
     */
    getSummary() {
        if (this.dataList.length === 0) {
            return {
                data_count: 0,
                data_columns: []
            };
        }

        const columns = Object.keys(this.dataList[0].row);
        return {
            data_count: this.dataList.length,
            data_columns: columns
        };
    }

    /**
     * Get total number of data points
     */
    getDataCount() {
        return this.dataList.length;
    }

    /**
     * Calculate quaternion multiplication (q1 * q2)
     */
    _multiplyQuaternions(q1, q2) {
        const result = quat.create();
        quat.multiply(result, q1, q2);
        return result;
    }

    /**
     * Convert quaternion to Euler angles (xyz convention: roll, pitch, yaw)
     * Returns angles in degrees
     */
    _quaternionToEuler(q) {
        const [x, y, z, w] = q;
        
        // Roll (x-axis rotation)
        const sinr_cosp = 2 * (w * x + y * z);
        const cosr_cosp = 1 - 2 * (x * x + y * y);
        const roll = Math.atan2(sinr_cosp, cosr_cosp);

        // Pitch (y-axis rotation)
        const sinp = 2 * (w * y - z * x);
        let pitch;
        if (Math.abs(sinp) >= 1) {
            pitch = Math.sign(sinp) * Math.PI / 2; // Use 90 degrees if out of range
        } else {
            pitch = Math.asin(sinp);
        }

        // Yaw (z-axis rotation)
        const siny_cosp = 2 * (w * z + x * y);
        const cosy_cosp = 1 - 2 * (y * y + z * z);
        const yaw = Math.atan2(siny_cosp, cosy_cosp);

        return {
            roll: roll * 180 / Math.PI,
            pitch: pitch * 180 / Math.PI,
            yaw: yaw * 180 / Math.PI
        };
    }

    /**
     * Get flight data at a specific index.
     * Returns dict with all relevant data including quaternions.
     */
    getDataAtIndex(index) {
        if (index < 0 || index >= this.dataList.length) {
            return null;
        }

        const data = this.dataList[index];
        const row = data.row;

        // Calculate quaternions
        // Helmet quaternion (corrected)
        const helmetCorrected = quat.fromValues(
            row.x_helmet || 0,
            row.y_helmet || 0,
            row.z_helmet || 0,
            row.w_helmet || 1
        );

        // Vehicle quaternion (world-relative)
        const vehicleWorld = quat.fromValues(
            row.x_vehicle || 0,
            row.y_vehicle || 0,
            row.z_vehicle || 0,
            row.w_vehicle || 1
        );

        // Helmet world quaternion = vehicle * helmet_corrected
        const helmetWorld = this._multiplyQuaternions(vehicleWorld, helmetCorrected);

        // Calculate ground speed
        const vvn = row.north || 0;
        const vve = row.east || 0;
        const vvd = row.down || 0;
        const gspeed = Math.sqrt(vvn * vvn + vve * vve + vvd * vvd);

        return {
            index: index,
            timestamp_seconds: data.timestampSeconds,
            timestamp_ns: data.timestampNs,

            // Vehicle quaternion (world-relative)
            VQX: vehicleWorld[0],
            VQY: vehicleWorld[1],
            VQZ: vehicleWorld[2],
            VQW: vehicleWorld[3],

            // Helmet quaternion (world-relative)
            HQX: helmetWorld[0],
            HQY: helmetWorld[1],
            HQZ: helmetWorld[2],
            HQW: helmetWorld[3],

            // Vehicle position and velocity
            VLAT: parseFloat(row.lat) || 0,
            VLON: parseFloat(row.lon) || 0,
            VALT: parseFloat(row.alt) || 0,
            VVN: parseFloat(vvn) || 0,
            VVE: parseFloat(vve) || 0,
            VVD: parseFloat(vvd) || 0,
            VINS: parseInt(row.mode) || 0,
            GSPEED: gspeed,
        };
    }

    /**
     * Binary search helper
     */
    _bisectLeft(arr, value) {
        let left = 0;
        let right = arr.length;
        while (left < right) {
            const mid = Math.floor((left + right) / 2);
            if (arr[mid] < value) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }

    /**
     * Find the data index closest to a given timestamp (in seconds).
     * Uses binary search for O(log n) performance.
     */
    findIndexForTimestamp(timestampSeconds) {
        if (this.dataList.length === 0) {
            return null;
        }

        const targetNs = timestampSeconds * 1e9;

        // Binary search for closest timestamp
        let idx = this._bisectLeft(this.timestampNsList, targetNs);

        // Handle edge cases
        if (idx === 0) {
            return 0;
        } else if (idx >= this.timestampNsList.length) {
            return this.timestampNsList.length - 1;
        }

        // Check which of the two adjacent points is closer
        const prevDiff = Math.abs(this.timestampNsList[idx - 1] - targetNs);
        const nextDiff = Math.abs(this.timestampNsList[idx] - targetNs);

        if (prevDiff < nextDiff) {
            return idx - 1;
        } else {
            return idx;
        }
    }

    /**
     * Efficiently extract all lat/lon/alt data.
     * Returns dict with 'lats', 'lons', 'alts' lists.
     */
    getAllPathData() {
        if (this.dataList.length === 0) {
            console.error('[getAllPathData] ERROR: dataList is empty!');
            return { lats: [], lons: [], alts: [] };
        }
        
        // Memory-efficient extraction: use pre-allocated arrays
        const length = this.dataList.length;
        const lats = new Array(length);
        const lons = new Array(length);
        const alts = new Array(length);
        
        // Extract in single pass
        for (let i = 0; i < length; i++) {
            const row = this.dataList[i].row;
            lats[i] = parseFloat(row.lat) || 0;
            lons[i] = parseFloat(row.lon) || 0;
            alts[i] = parseFloat(row.alt) || 0;
        }
        
        return {
            lats,
            lons,
            alts
        };
    }

    /**
     * Efficiently extract all attitude data (yaw, pitch, roll) from vehicle quaternions.
     * Returns dict with 'yaws', 'pitches', 'rolls' lists (in degrees).
     */
    getAllAttitudeData() {
        const length = this.dataList.length;
        const yaws = new Array(length);
        const pitches = new Array(length);
        const rolls = new Array(length);

        if (length === 0) {
            console.error('[getAllAttitudeData] ERROR: dataList is empty');
            return { yaws: [], pitches: [], rolls: [] };
        }

        // Check first row for quaternion columns
        const firstRow = this.dataList[0].row;
        const hasVehicleQuat = firstRow.x_vehicle !== undefined && firstRow.x_vehicle !== null;
        
        if (!hasVehicleQuat) {
            console.error('[getAllAttitudeData] ERROR: No vehicle quaternion data found');
            return { yaws: [], pitches: [], rolls: [] };
        }

        // Memory-efficient extraction: pre-allocated arrays, single pass
        let errorCount = 0;
        for (let i = 0; i < length; i++) {
            const row = this.dataList[i].row;
            
            try {
                const vehicleWorld = quat.fromValues(
                    parseFloat(row.x_vehicle) || 0,
                    parseFloat(row.y_vehicle) || 0,
                    parseFloat(row.z_vehicle) || 0,
                    parseFloat(row.w_vehicle) || 1
                );

                const euler = this._quaternionToEuler(vehicleWorld);
                rolls[i] = euler.roll;
                pitches[i] = euler.pitch;
                yaws[i] = euler.yaw;
            } catch (error) {
                errorCount++;
                // Push default values
                rolls[i] = 0;
                pitches[i] = 0;
                yaws[i] = 0;
            }
        }

        if (errorCount > 0 && errorCount <= 10) {
            console.warn(`[getAllAttitudeData] ${errorCount} errors during extraction`);
        }

        return {
            yaws,
            pitches,
            rolls
        };
    }
}

module.exports = FlightDataProcessor;

